ProjectSource/app-library/src/test/java/com/kayako/sdk/android/k5/kre/base/kase/KreCaseSubscriptionTest.java

```diff
@@ -1,9 +1,307 @@
+
+        //Act
+        kreCaseSubscription.subscribe(kreCredentials, channel, isAgent, onSubscriptionListener);
+
+        //Assert
+        final boolean mIsAgent = Whitebox.getInternalState(kreCaseSubscription, "mIsAgent");
+        final List<KreSubscription.OnSubscriptionListener> mChildListeners =
+                    Whitebox.getInternalState(kreCaseSubscription, "mChildListeners");
+        errorCollector.checkThat(mChildListeners.size(), is(1));
+        errorCollector.checkThat(mIsAgent, is(isAgent));
+    }
+
+    @Test
+    public void unSubscribe() {
+        //Arrange
+        final String channel = "mode";
+        final boolean isAgent = true;
+        suppress(method(KreSubscription.class, "runUnSubscribeTask"));
+        kreCaseSubscription.subscribe(kreCredentials, channel, isAgent, onSubscriptionListener);
+
+        //Act
+        kreCaseSubscription.unSubscribe(onSubscriptionListener);
+
+        //Assert
+        final List<KreSubscription.OnSubscriptionListener> mChildListeners =
+                Whitebox.getInternalState(kreCaseSubscription, "mChildListeners");
+        errorCollector.checkThat(mChildListeners.size(), is(0));
+    }
+
+    @Test
+    public void unSubscribeWithResetVariables() {
+        //Arrange
+        suppress(method(KreSubscription.class, "runUnSubscribeTask"));
+        Whitebox.setInternalState(kreCaseSubscription, "mChildListeners", mChildListeners);
+        when(mChildListeners.size()).thenReturn(1);
+
+        //Act
+        kreCaseSubscription.unSubscribe(onSubscriptionListener);
+
+        //Assert
+        final KreSubscription.OnSubscriptionListener mMainListenerLocal =
+                Whitebox.getInternalState(kreCaseSubscription, "mMainListener");
+        final List<MinimalClientTypingListener> mMinimalClientTypingListenersList =
+                Whitebox.getInternalState(kreCaseSubscription, "mMinimalClientTypingListeners");
+        final List<RawClientTypingListener> mClientTypingListenersList =
+                Whitebox.getInternalState(kreCaseSubscription, "mClientTypingListeners");
+        final List<RawClientActivityListener> mClientActivityListenersList =
+                Whitebox.getInternalState(kreCaseSubscription, "mClientActivityListeners");
+        final List<RawCaseChangeListener> mCaseChangeListenersList =
+                Whitebox.getInternalState(kreCaseSubscription, "mCaseChangeListeners");
+        final List<RawUserOnCasePresenceListener> mUserPresenceListenersList =
+                Whitebox.getInternalState(kreCaseSubscription, "mUserPresenceListeners");
+        final List<RawCasePostChangeListener> mRawCasePostChangeListenersList =
+                Whitebox.getInternalState(kreCaseSubscription, "mRawCasePostChangeListeners");
+        errorCollector.checkThat(mMainListenerLocal, nullValue());
+        errorCollector.checkThat(mMinimalClientTypingListenersList.size(), is(0));
+        errorCollector.checkThat(mClientTypingListenersList.size(), is(0));
```

Comment:
⚠️  Here, the code is using `Collection.size()` for testing emptiness. Although, using `Collection.size()` to test for emptiness works, using `Collection.isEmpty()` makes the code more readable and can be more efficient. The time complexity of any `isEmpty()` method implementation should be **O(1)** whereas some implementations of `size()` can be **O(n)**.  

**[More details](https://confluence.devfactory.com/x/uyCkEw)**

## The solution in this case

```JAVA
errorCollector.checkThat(mClientTypingListenersList.isEmpty(), is(true));
```
 

<a id="cf943bc0-4980-11e8-86d4-9358f2b1b645" href="none" codereview-uuid-reserved-tag />